### PR TITLE
feat(drivers): Driver for MAX7318 GPIO expander

### PR DESCRIPTION
**Pull request by zhiayang**
Adds a driver for the [MAX7318](https://datasheets.maximintegrated.com/en/ds/MAX7318.pdf) 16-port I2C GPIO expander. Doesn't support the interrupt feature (yet?).

Has been tested on real hardware.

I've moved the existing mcp23017 driver to its own subfolder as well to keep things a little more organised.

### CODE REVIEW

File: app/drivers/gpio/max7318/Kconfig [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# MAX7318 GPIO configuration options
+
+# Copyright (c) 2022 The ZMK Contributors
+# SPDX-License-Identifier: MIT
+
+menuconfig GPIO_MAX7318
+	bool "MAX7318 I2C-based GPIO chip"
+	depends on I2C
+	select HAS_DTS_GPIO
```
Comment (petejohanson) [inline app/drivers/gpio/max7318/Kconfig:9]: See the 595 gpio driver I just merged for an example of how to auto-enable this if found in the DT.

File: app/drivers/gpio/max7318/gpio_max7318.c [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+    LOG_INF("device initialised at 0x%x (i2c=%s)", config->device_addr, config->i2c_dev_name);
+
+    k_sem_init(&drv_data->lock, 1, 1);
+    return 0;
+}
+
+#define MAX7318_INIT(inst)                                                                         \
+    static struct max7318_config max7318_##inst##_config = {                                       \
+        .i2c_dev_name = DT_INST_BUS_LABEL(inst),                                                   \
```
Comment (petejohanson) [inline app/drivers/gpio/max7318/gpio_max7318.c:295]: The common stuff should be initialized here, see https://github.com/zmkfirmware/zmk/blob/main/app/drivers/gpio/gpio_595.c#L200 for an example.

Also, Zephyr has newer APIs for the bus stuff, which isn't incorporated into our local mcp23017 driver yet, but should be used for newer drivers. The I2C idea can be seen at https://github.com/zmkfirmware/zmk/blob/main/app/drivers/gpio/gpio_595.c#L205 but there is an equivalent macro and API for SPI.

File: app/drivers/gpio/max7318/gpio_max7318.c [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+static int write_registers(const struct device *dev, uint8_t reg, uint16_t value) {
+    struct max7318_drv_data *const drv_data = (struct max7318_drv_data *const)dev->data;
+
+    LOG_DBG("max7318: write: reg[0x%X] = 0x%X, reg[0x%X] = 0x%X", reg, (value & 0xFF), (reg + 1),
+            (value >> 8));
+
+    uint16_t data = sys_cpu_to_le16(value);
+    uint16_t dev_addr = ((struct max7318_config *)dev->config)->device_addr;
+
+    return i2c_burst_write(drv_data->i2c, dev_addr, reg, (uint8_t *)&data, sizeof(data));
```
Comment (petejohanson) [inline app/drivers/gpio/max7318/gpio_max7318.c:82]: Move to the `_dt` version of this.

File: app/drivers/gpio/max7318/gpio_max7318.c [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+ * @param buf   Buffer to read data into.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int read_registers(const struct device *dev, uint8_t reg, uint16_t *buf) {
+    struct max7318_drv_data *const drv_data = (struct max7318_drv_data *const)dev->data;
+    uint16_t dev_addr = ((struct max7318_config *)dev->config)->device_addr;
+
+    uint16_t data = 0;
+    int ret = i2c_burst_read(drv_data->i2c, dev_addr, reg, (uint8_t *)&data, sizeof(data));
```
Comment (petejohanson) [inline app/drivers/gpio/max7318/gpio_max7318.c:47]: Ditto, move to the `_dt` version of this.

**Review by petejohanson — CHANGES_REQUESTED** [review]
Thanks for working on this!

I don't really agree w/ the need for separate dirs yet for this, IMHO, can you please not mix that in w/ the new driver here?

Other than that, a few suggestions, I imagine mostly stemming from you copying our current out of tree driver that's missing some upstream QOL improvements.

**Comment by zhiayang** [comment]
Addressed the PR comments

File: app/drivers/gpio/max7318/Kconfig (hunk 1 above)
Comment (zhiayang) [reply app/drivers/gpio/max7318/Kconfig]: yep, copied it over.

File: app/drivers/gpio/max7318/gpio_max7318.c (hunk 2 above)
Comment (zhiayang) [reply app/drivers/gpio/max7318/gpio_max7318.c]: I think you meant `s/spi/i2c/`. I hope I made the changes correctly (it compiles); will test with my board to make sure it works when I get the time (soon)

File: app/drivers/gpio/max7318/gpio_max7318.c (hunk 3 above)
Comment (zhiayang) [reply app/drivers/gpio/max7318/gpio_max7318.c]: Done.

File: app/drivers/gpio/max7318/gpio_max7318.c (hunk 4 above)
Comment (zhiayang) [reply app/drivers/gpio/max7318/gpio_max7318.c]: Done.

**Comment by zhiayang** [comment]
@petejohanson are there any more issues preventing this from being merged? let me know what else to change. thanks!

File: app/drivers/gpio/Kconfig.max7318 [hunk 5]
```diff
@@ -0,0 +1,25 @@
+# MAX7318 GPIO configuration options
+
+# Copyright (c) 2022 The ZMK Contributors
+# SPDX-License-Identifier: MIT
+
+DT_COMPAT_ZMK_GPIO_MAX7318 := zmk,gpio-max7318
```
Comment (petejohanson) [inline app/drivers/gpio/Kconfig.max7318:6]: This doesn't match the compatible defined in the yaml binding or used in the code. Accident or intentional?

**Review by petejohanson — COMMENTED** [review]
Just one minor nitpick, probably keeping this from automatically being enabled if the DTS contains a compatible node.

File: app/drivers/gpio/Kconfig.max7318 (hunk 5 above)
Comment (zhiayang) [reply app/drivers/gpio/Kconfig.max7318:6]: Ah, my bad. The MCP23017 driver doesn't have `gpio-` in front, but the 595 does -- which one should I follow here?

Thanks!

Comment (petejohanson) [reply app/drivers/gpio/Kconfig.max7318:6]: Follow the mcp format is fine.

**Comment by zhiayang** [comment]
should've fixed them now!

File: app/drivers/gpio/Kconfig.max7318 [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# MAX7318 GPIO configuration options
+
+# Copyright (c) 2022 The ZMK Contributors
+# SPDX-License-Identifier: MIT
+
+DT_COMPAT_ZMK_GPIO_MAX7318 := maxim,max7318
```
Comment (petejohanson) [inline app/drivers/gpio/Kconfig.max7318:6]:
```suggestion
DT_COMPAT_MAXIM_MAX7318 := maxim,max7318
```

File: app/drivers/gpio/Kconfig.max7318 [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+menuconfig GPIO_MAX7318
+	bool "MAX7318 I2C-based GPIO chip"
+	default $(dt_compat_enabled,$(DT_COMPAT_ZMK_GPIO_MAX7318))
```
Comment (petejohanson) [inline app/drivers/gpio/Kconfig.max7318:10]:
```suggestion
	default $(dt_compat_enabled,$(DT_COMPAT_MAXIM_MAX7318))
```

**Review by petejohanson — CHANGES_REQUESTED** [review]
Just one other nitpick for the naming changes.

**Comment by zhiayang** [comment]
done!

**Review by petejohanson — APPROVED** [review]
Thanks!

**Pull request merged**